internal/cli/command/controlplane/debug.go

```diff
@@ -0,0 +1,241 @@
+
+	return cmd
+}
+
+type debugMetadata struct {
+	Timestamp       time.Time
+	Host            host.InfoStat
+	CLIVersion      string
+	WorkspacePath   string
+	Workspaces      []string
+	InstallerImages []string
+	DockerVersion   string
+}
+
+func runDebug(options debugOptions, banzaiCli cli.Cli) error {
+	logBuffer := new(bytes.Buffer)
+	log.SetOutput(io.MultiWriter(os.Stderr, logBuffer))
+
+	if err := options.Init(); err != nil {
+		return err
+	}
+
+	path := options.filepath()
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.WrapIff(err, "failed to create archive at %q", path)
+	}
+	defer f.Close()
+
+	gzWriter := gzip.NewWriter(f)
+	defer gzWriter.Close()
+
+	tarWriter := tar.NewWriter(gzWriter)
+	defer tarWriter.Close()
+
+	addDir(tarWriter, "")
+
+	meta := debugMetadata{
+		Timestamp:       time.Now(),
+		CLIVersion:      banzaiCli.Version(),
+		WorkspacePath:   options.workspace,
+		Workspaces:      strings.Split(simpleCommand("ls ~/.banzai/pipeline"), "\n"),
+		InstallerImages: strings.Split(simpleCommand("docker image list|grep pipeline-installer"), "\n"),
+		DockerVersion:   simpleCommand("docker --version"),
+	}
+
+	if info, err := host.Info(); err == nil {
+		meta.Host = *info
+	}
+
+	bytes, _ := yaml.Marshal(meta)
+	logResult("add meta.yaml", addFile(tarWriter, "meta.yaml", string(bytes)))
+
+	logResult("create pipeline/", addDir(tarWriter, "pipeline"))
+	logResult("add values.yaml", copyFile(tarWriter, "pipeline/values.yaml", options.valuesPath()))
+
+	// run some terraform diagnostics commands to catch their output in the logs folder
+	var values map[string]interface{}
+	logResult("read values", options.readValues(&values))
+	_, env, err := getImageMetadata(options.cpContext, values, true)
+	logResult("get image metadata", err)
+	logResult("run tf plan", runTerraform("plan", options.cpContext, env))
+	logResult("run tf graph", runTerraform("graph", options.cpContext, env))
+	logResult("run tf state list", runTerraform("state list", options.cpContext, env))
+
+	logResult("create pipeline/installer-logs", addDir(tarWriter, "pipeline/installer-logs"))
+	logDir, logFiles, err := options.listLogs()
+	if err != nil {
+		log.Errorf("listing log files failed: %v", err)
+	} else {
+		for _, file := range logFiles {
+			logResult("add log file", copyFile(tarWriter, filepath.Join("pipeline/installer-logs", file), filepath.Join(logDir, file)))
+		}
+	}
+
+	logResult("add pipeline/files.txt", addFile(tarWriter, "pipeline/files.txt", simpleCommand("find", options.workspace, "-ls")))
+
+	logResult("create pipeline/resources", addDir(tarWriter, "pipeline/resources"))
+	logResult("add pipeline/resources/all.txt", addFile(tarWriter, "pipeline/resources/all.txt", fetchContainerCommandOutputAndError(options.cpContext, []string{"kubectl", "get", "all", "-A", "-owide"}, env)))
```

Comment:
all.txt is a text output for a quick look on the whole cluster. resources that are explicitly interesting for us, are requested elsewhere.

Reply:
I know and see, but I find it tedious to maintain and hard to overview easily, that's why I mentioned the exhaustive approach I thought about.

Reply:
As an alternative suggestion to my original idea I would look for ways to extract "templates" for similar executions in order to write this code block in a more declarative style where the important information - such as file path/resource kind - is defined on one abstraction level and all the redundant function calls and execution details are on another level.

For example all the `kubectl get ${RESOURCE_KIND} --output yaml` operations could be extracted and a list of resource kinds could be given to a function which executes the same for all kinds.

(Basically I would give a layer more hierarchy to the run-command|add-log-to-bundle|log-result calls block, because this way for me the meaningful details seem to be lost among the redundant "plumbing".)

What I imagine is something like this:
```golang
directories := []string{ // Note: I would even propose directories should be derived from the file paths, so this is optional if we want to keep them for some reason.
    "",
    "pipeline",
    ...
}
files := []BundleFile{
    {
        Operation: BundleOperationAddFile,
        Path: "meta.yaml",
        Source: bytes,
    },
    {
        Operation: BundleOperationCopyFile,
        Path: "pipeline/values.yaml",
        Source: options.valuesPath(),
    },
    {
        Operation: BundleOperationAddFile,
        Path: "pipeline/resources/all.txt",
        Source: []string{"kubectl", "get", "all", "-A", "-owide"},
    },
}
AssembleBundle(directories, files)
```

Where Source is an `interface{}` which can be type switched into
- `string` which if a file path + operation is copy then the file is copied, otherwise the string is written to the file as content.
- `[]byte` which is written as raw content.
- `[]string` which is interpreted as a command and is executed according to the corresponding command handler (`fetchContainerCommandOutput` for `kubectl`, etc., `simpleCommand` by default).

This doesn't cover special cases such as terraform executions which are very custom.
I haven't given this much consideration, just a couple minutes and just wrote what came to try and convey what I meant, it is not a merge-ready solution I know, it's a conversation starter which I'm glad to develop further if we see potential in it.
But I gotta go now.

Reply:
(This is related to [the other thread](https://github.com/banzaicloud/banzai-cli/pull/297#discussion_r481997174).)

Reply:
This part is now something easier to read.